### PR TITLE
Adds rendering of ALL grayscale LD spectrograms to standard indices generation

### PR DESCRIPTION
**Pull request by towsey**
I have finished with issue #199. I combined the grey scale spectrograms into one image because it is most helpful to be able to compare them side by side. If spectrograms are too wide then they are joined vertically.

sm304256_0 1_20151114_001652__keys

### CODE REVIEW

File: src/AudioAnalysisTools/DSP/FrequencyScale.cs [hunk 1]
```diff
@@ -396,9 +396,16 @@ public static void DrawFrequencyLinesOnImage(Bitmap bmp, int[,] gridLineLocation
             for (int b = 0; b < bandCount; b++)
             {
                 int y = height - gridLineLocations[b, 0];
+                int hertzValue = gridLineLocations[b, 1];
+                if (bmp.Width < 120)
+                {
+                    // convert label to kHz so does not hide too much spectrogram
+                    hertzValue = hertzValue / 1000;
+                }
+
```
Comment (atruskie) [inline src/AudioAnalysisTools/DSP/FrequencyScale.cs]: Please revert this change.

We aim for consistency in the images we output. If an image is too short then it is not well suited to a LD spectrogram.

Either all values on all images are in Hz, or in KHz, not both

File: src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramRGB.cs [hunk 2]
```diff
@@ -499,23 +506,21 @@ public void LoadSpectrogramDictionary(Dictionary<string, double[,]> dictionary)
...
+            // init List of track images
+            var listOfImages = new List<Image>();
+
             foreach (string key in keys)
             {
+                if (key == "SUM" || key == "DIF")
+                {
+                    continue;
+                }
+
```
Comment (atruskie) [inline src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramRGB.cs]: Why not use the `DoDisplay` value from IndexPropertiesConfig?

File: src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramRGB.cs [hunk 3]
```diff
@@ -540,10 +545,38 @@ public void DrawGreyScaleSpectrograms(DirectoryInfo opdir, string opFileName, st
...
+            // check how wide combined image will be. If tracks are wider than 180 = 3 hours, then go vertical
+            if (listOfImages[0].Width * imageCount > 180 * imageCount)
+            {
+                var combinedImage = ImageTools.CombineImagesVertically(listOfImages);
+                combinedImage?.Save(path);
+            }
+            else
+            {
+                var combinedImage = ImageTools.CombineImagesInLine(listOfImages);
+                combinedImage?.Save(path);
```
Comment (atruskie) [inline src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramRGB.cs]: This is not OK. We never want to generate inconsistent output.

Our images always should have the same format and structure, otherwise, it is hard to script and use the results.

File: src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramRGB.cs [hunk 4]
```diff
@@ -540,10 +545,38 @@ public void DrawGreyScaleSpectrograms(DirectoryInfo opdir, string opFileName, st
...
+
+                // save the image - the directory for the path must exist
+                //var path = FilenameHelpers.AnalysisResultPath(opdir, opFileName, key, "png");
+                //indexImage?.Save(path);
+
+                listOfImages.Add(indexImage);
+            }
+
+            // save the combined track image - the directory for the path must exist
+            var path = FilenameHelpers.AnalysisResultPath(opdir, opFileName, "KEYS", "png");
```
Comment (atruskie) [inline src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramRGB.cs]: This change breaks existing functionality. Output each image, separately, please.

If you want to output a combined image, it needs to hidden behind a config setting that is not enabled by default.

**Comment by atruskie** [comment]
@towsey, I fixed the unit tests but uncovered a previously unseen bug.

See the commit details of ec2e73e and this code:

https://github.com/QutEcoacoustics/audio-analysis/blob/ec2e73e0eee680ec670d66334e470421d169f304/src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramRGB.cs#L499-L506

Is that change acceptable?

**Comment by towsey** [comment]
Hi Anthony,

Yes your proposed fix looks a good solution to me. Cheers, Michael


Dr. Michael Towsey
ORCID iD http://orcid.org/0000-0002-8246-7151
QUT Ecoacoustics Research Group  |  www.Ecosounds.org<http://www.ecosounds.org/>
Science and Engineering Faculty
S block, Level 10, Room 1002, Garden Point Campus,
Queensland University of Technology,
Brisbane, Queensland 4001
Australia
Phone +61 7 3138 9381
Fax     +61 7 3138 9390
________________________________
